chairman/Chairman.hs

```diff
@@ -0,0 +1,434 @@
+              coreNodeId
+              chainsVar
+              securityParam
+              maxBlockNo
+              (contramap show tracer)
+              nullTracer
+              nullTracer
+              pInfoConfig)
+            Nothing
+            (localSocketAddrInfo (localSocketFilePath coreNodeId))
+
+
+data ChairmanTrace blk
+  = NotFoundCommonBlock [Maybe blk]
+  | WitnessedConsensusAt SlotNo [Point blk]
+
+instance (Condense blk, Condense (HeaderHash blk)) => Show (ChairmanTrace blk) where
+    show (NotFoundCommonBlock blks)
+      = "not found common block" ++ condense blks
+    show (WitnessedConsensusAt slotNo tips)
+      = mconcat
+      [ "witnessed consensus at slot "
+      , condense slotNo
+      , " current tips: "
+      , condense tips
+      ]
+
+
+--
+-- Shared State, and its API.
+--
+
+
+-- | Shared state between chain-sync clients.  Each chain-sync client will write to the
+-- corresponding entry.
+--
+type ChainsVar m blk = TVar m (Map CoreNodeId (ChainFragment blk))
+
+
+-- | Add a single block to the chain.
+--
+addBlock
+    :: forall blk m.
+       ( MonadSTM m
+       , HasHeader blk
+       )
+    => CoreNodeId
+    -> ChainsVar m blk
+    -> blk
+    -> STM m ()
+addBlock coreNodeId chainsVar blk =
+    modifyTVar' chainsVar (Map.adjust (CF.addBlock blk) coreNodeId)
+
+
+data ChairmanError blk = NodeMisconduct [blk]
+
+instance Condense blk => Show (ChairmanError blk) where
+    show (NodeMisconduct blks) = "NodeMisconduct " ++ condense blks
+
+instance (Condense blk, Typeable blk) => Exception (ChairmanError blk)
+
+
+-- | Check if the oldest immutable tip agree, keep forks from the last common
+-- block (including it).
+--
+checkAndSubtract
+    :: forall blk m.
+       ( MonadSTM m
+       , HasHeader blk
+       , Condense blk
+       )
+    => ChainsVar m blk
+    -> Int
+    -> STM m (ChairmanTrace blk)
+checkAndSubtract chainsVar securityParam = do
+    chains <- readTVar chainsVar
+    let (res, chains') = subtractPure chains
+    writeTVar chainsVar $! chains'
+    return res
+  where
+    -- it isn't really pure as it throws the 'NodeMisconduct' exception
+    subtractPure :: Map CoreNodeId (ChainFragment blk)
+                 -> ( ChairmanTrace blk
+                    , Map CoreNodeId (ChainFragment blk)
+                    )
+    subtractPure chains =
+      let tips :: [Maybe blk]
+          tips = foldl' (\tips' cf -> CF.head cf : tips') [] chains
+
+          -- find intersection of all the chains; the CLI guarantees that
+          -- there's at least one entry in @chains@, thus @fold1@ is safe here.
+          common :: ChainFragment blk
+          common =
+            foldl1
+              (\common' cf ->
+                case CF.intersectChainFragments common' cf of
+                  Nothing -> CF.Empty
+                  Just (_, common'', _, _)
+                          -> common'')
+              chains
+
+          mbHeadSlot = CF.headSlot common
```

Comment:
`mbCommonHeadSlot`?

See the next comment: use a `Point` instead!

Reply:
> See the next comment: use a Point instead!

I don't see it; Either will work, because the slot is occupied by the same block in each chain.  But using `Point` might be more instructive.